src/app/categories/[categoryId]/page.tsx

```diff
@@ -0,0 +1,79 @@
+import { HEADER_NAV } from "@/Constants";
+import Header from "@/components/Header/Header";
+import ProductCard from "@/components/ProductCard/ProductCard";
+import { getCategory } from "@/model/CategoryModel";
+import { getProductsByCategory } from "@/model/ProductModel";
+import { Product } from "@/services/ProductAPI";
+import { Metadata } from "next";
+import { notFound } from "next/navigation";
+
+export async function generateMetadata({
+  params,
+}: {
+  params: { categoryId: string };
+}): Promise<Metadata> {
+  let category = await getCategory(params.categoryId);
+
+  if (category === undefined) {
+    return {
+      title: "Not found!",
+    };
+  }
+  return {
+    title: category.name,
+  };
+}
+
+export default async function Page({
+  params,
+}: {
+  params: { categoryId: string };
+}) {
+  let category = await getCategory(params.categoryId);
+  if (category === undefined) {
+    return notFound();
+  }
+  let products: Product[];
+  try {
+    products = await getProductsByCategory(params.categoryId);
+    if (products.length === 0) {
+      return (
+        <div>
+          <Header nav={HEADER_NAV} />
+          <p className="font-bold text-center text-xl"> {category.name}</p>
+          <main>
+            <p className="m-2 text-center">
+              There are no products for this category
+            </p>
+          </main>
+        </div>
+      );
+    }
+  } catch (err) {
+    console.log("Failed to select products", err);
+    return (
+      <div>
+        <Header nav={HEADER_NAV} />
+        <main>
+          <p className="m-2 text-center"> Failed to list products </p>
+        </main>
+      </div>
+    );
+  }
+
+  return (
+    <div>
+      <Header nav={HEADER_NAV} />
+      <p className="font-bold text-center text-xl"> {category.name}</p>
+      <div className="grid card-grid-cols-3 mt-4 gap-4">
+        {products.map((product) => {
+          return (
+            <div key={product.id}>
+              <ProductCard product={product} />
+            </div>
+          );
+        })}
+      </div>
+    </div>
```

Comment:
We have a LOT of duplicate code. We have the header 3 times. Let's create a new component and we'll use it like this.

```tsx
<CategoryPage>
   <p className="m-2 text-center">
     There are no products for this category
   </p>
</CategoryPage>
```

```tsx
<CategoryPage>
  <p className="m-2 text-center"> Failed to list products </p>
</CategoryPage>
```

or

```tsx
<CategoryPage>
      <div className="grid card-grid-cols-3 mt-4 gap-4">
        {products.map((product) => {
          return (
            <div key={product.id}>
              <ProductCard product={product} />
            </div>
          );
        })}
      </div>
</CategoryPage>
```

NOTE: you have to figure out what props this component needs to get. The hint I give you is how we want to use it, to remove code duplication.